### PR TITLE
[msbuild] Do not allow iOS and Mac to share the same Xamarin.MacDev* …

### PR DESCRIPTION
**Pull request by jstedfast**
…assemblies

Fixes https://bugzilla.xamarin.com/show_bug.cgi?id=52729

### CODE REVIEW

**Comment by jstedfast** [comment]
@rolfbjarne I need your help with fixing up the Makefile to correctly copy the Xamarin.iOSDev.Tasks* assemblies from their proper locations.

The problem is that the Makefile rules use the assembly name to construct the path, e.g.:

Xamarin.iOSDev.Tasks -> Xamarin.iOSDev.Tasks/bin/Release/Xamarin.iOSDev.Tasks.dll

but this assembly is actually located in Xamarin.MacDev.Tasks/bin/Release/Xamarin.iOSDev.Tasks.dll

**Comment by jstedfast** [comment]
@rolfbjarne I thought of that, but that does make sharing code between Mac & iOS a bit harder...

**Comment by rolfbjarne** [comment]
@jstedfast why? can't you just link to the source files in the corresponding projects?

**Comment by jstedfast** [comment]
iOS.Tasks.Core subclasses MyTaskBase and calls it MyTaskBase and then iOS.Tasks subclasses again and calls it MyTask

**Comment by jstedfast** [comment]
Yes, this is all safe syntactically from a namespace point of view all in the same assembly, but it's still awkward

**Comment by jstedfast** [comment]
anyway, I found a way to make fix the Makefile in a simple way, so all works now :)

**Comment by jstedfast** [comment]
I don't understand the Cecil approach or how that could/would work.

File: msbuild/Makefile [hunk 1]
```diff
@@ -326,17 +326,22 @@ $(foreach dll,$(MSBUILD_TASK_ASSEMBLIES),$(dll)/bin/$(CONFIG)/$(dll).dll): .buil
 .PHONY: $(MSBUILD_SYMLINKS)
 
 # Xamarin.MacDev.dll comes from the Xamarin.MacDev repository
-# Newtonsoft.Json.dll comes from a NuGet
```
Comment (spouliot) [inline msbuild/Makefile]: Can you provide QA a list of tasks that required `Newtonsoft.Json.dll` so they can make sure the changes are tested correctly ?

File: msbuild/Xamarin.MacDev.Tasks/Xamarin.iOSDev.Tasks.csproj [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    <Reference Include="System" />
+    <Reference Include="Microsoft.Build.Engine" />
+    <Reference Include="Microsoft.Build.Framework" />
+    <Reference Include="Microsoft.Build.Tasks.v4.0" />
+    <Reference Include="Microsoft.Build.Utilities.v4.0" />
+    <Reference Include="System.Core" />
+    <Reference Include="System.Xml" />
+    <Reference Include="Mono.Cecil, Version=0.9.5.0, Culture=neutral, PublicKeyToken=0738eb9f132ed756">
+      <Private>True</Private>
+      <HintPath Condition=" Exists('..\..\packages\Mono.Cecil.0.9.5.0\lib\net40\Mono.Cecil.dll') ">..\..\packages\Mono.Cecil.0.9.5.0\lib\net40\Mono.Cecil.dll</HintPath>
```
Comment (spouliot) [inline msbuild/Xamarin.MacDev.Tasks/Xamarin.iOSDev.Tasks.csproj]: that's also an assembly that can change between versions
and can include breaking changes (it's happening with mono 5.8)

**Comment by kzu** [comment]
Attached is the basic idea @jstedfast with three fake projects:
1. MacDev (the one upstream in Xamarin.MacDev), the one you want to reuse the *code* but not the assembly name
2. Two tasks projects, MacDev.Tasks and iOSDev.Tasks which both reference 1.

Both tasks project reference the MacDev one like so:

```xml
    <ProjectReference Include="..\Xamarin.MacDev\Xamarin.MacDev.csproj">
      <Project>{46e80364-c523-4b95-8c2b-ac1f1d047a15}</Project>
      <AdditionalProperties>AssemblyName=Xamarin.iOSDev;OutputPath=$(MSBuildProjectDirectory)$(OutputPath)</AdditionalProperties>
      <Name>Xamarin.MacDev</Name>
    </ProjectReference>
```

Passing the desired `AssemblyName` as an additional property to pass to the project reference when building it. The `MacDev` project in turn just needs to honor those overrides when present:

```xml
    <AssemblyName Condition="'$(AssemblyName)' == ''">Xamarin.MacDev</AssemblyName>
    ...
    <OutputPath Condition="'$(Condition)' == ''">bin\Release\</OutputPath>
```

This means you don't need to maintain duplicate projects at all.

[Xamarin.MacDev.zip](https://github.com/xamarin/xamarin-macios/files/1433150/Xamarin.MacDev.zip)

**Comment by kzu** [comment]
I have to say though, that using the [ILRepack nuget](https://www.nuget.org/packages/ILRepack) package and [a bit of MSBuild](https://github.com/NuGet/NuGet.Build.Packaging/blob/dev/src/Build/NuGet.Build.Packaging.Tasks/NuGet.Build.Packaging.Tasks.targets#L94-L128) sounds even better, simpler and more fool proof.

**Comment by jstedfast** [comment]
ILRepack is licensed under the Apache 2.0 license and is not MIT, is that something we can use?

**Comment by jstedfast** [comment]
This PR has been replaced by https://github.com/xamarin/xamarin-macios/pull/3007

**Pull request closed without merging**